[RDY] pygccxml + castxml + pybindgen

teto opened the pull request:
###### Motivation for this change
I use the simulator www.nsnam.org in my work. I need these libraries in order to generate the python bindings castxml + pygccxml. 

Castxml linking stage fails because it can't find -ltinfo. Comments in the repository say it's not available. Not sure what to do. 

###### Things done


<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

peti commented on pkgs/development/compilers/ghc/7.4.2-binary.nix:
```diff
@@ -34,7 +34,7 @@ stdenv.mkDerivation rec {
     # during linking
     stdenv.lib.optionalString stdenv.isDarwin ''
-      export NIX_LDFLAGS+=" -no_dtrace_dof"
+      export NIX_LDFLgGS+=" -no_dtrace_dof"
```
Huh? That's probably not intentional?

teto replied:
I switch between 3 different keyboards a day and sometimes I make mistakes. Will fix

peti commented on lib/licenses.nix:
```diff
@@ -39,6 +39,11 @@ lib.mapAttrs (n: v: v // { shortName = n; }) rec {
   };
 
+  apache2 = spdx {
+    spdxId = "AL-2.0";
```
According to https://spdx.org/licenses/, that license does not exist.

teto replied:
Didn't know spdxId was an official register so I took a random one. Apache 2 license is registered though https://spdx.org/licenses/Apache-2.0.html. I will fix it

peti commented on pkgs/top-level/python-packages.nix:
```diff
@@ -1221,7 +1221,7 @@ in {
 
   batinfo = buildPythonPackage rec {
-    version = "0.3";
+    version = "-1.3";
```
This is not right; the version must start with a digit.

teto replied:
typically typo mistake too. I will mark the PR as RFC once I make sure everything works. My last commit "Now works" was a shout of joy after sphalerite helped me solve sthg I 'd spent quite some time. thanks for the review

teto commented:
tests fail because of an undesired warning generated by `-Wunused-command-line-argument`. Is that possible that the flag was added by cmake ? I looked at the cmake builder and could not find it. Might it be in clang-wrapped? tried in a nix-shell --pure with the same results. More details here
https://github.com/CastXML/CastXML/issues/90#issuecomment-337270818

FRidh commented on pkgs/development/python-modules/pygccxml/default.nix:
```diff
@@ -0,0 +1,26 @@
+
+  propagatedBuildInputs = [ pkgs.castxml ];
+
+  doCheck = false;
```
can you include a comment with a reason

FRidh commented on lib/licenses.nix:
```diff
@@ -39,6 +39,11 @@ lib.mapAttrs (n: v: v // { shortName = n; }) rec {
     url = http://developer.amd.com/amd-license-agreement/;
   };
 
+  apache2 = spdx {
```
This is the `asl20`

FRidh commented on pkgs/development/tools/castxml/default.nix:
```diff
@@ -0,0 +1,70 @@
+{ pkgs
```
do not pass in `pkgs`

teto committed:
castxml: init at 20180403

4 tests are ignored because the regex don't match with custom clang
output see:
https://github.com/CastXML/CastXML/issues/90

teto committed:
pygccxml: init at 1.9.1

Python library to play with castxml output, aka your program's clang-produced AST.

teto commented:
I am trying to enable pygccxxml tests (previously disabled).

So first it generated
```
[nix-shell:~/pygccxml-v1.9.1-src]$ python -m unittests.test_all/home/teto/pygccxml-v1.9.1-src/unittests/data/core_types.hpp:9:10: fatal error: 
      'iostream' file not found
#include <iostream>
         ^~~~~~~~~~
1 error generated.
Traceback (most recent call last):
  File "/nix/store/2lb6nn8ivk1alhckv43n7734lqwbw7h9-python3-3.6.2/lib/python3.6/runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "/nix/store/2lb6nn8ivk1alhckv43n7734lqwbw7h9-python3-3.6.2/lib/python3.6/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/home/teto/pygccxml-v1.9.1-src/unittests/test_all.py", line 193, in <module>
    sys.exit(run_suite())
  File "/home/teto/pygccxml-v1.9.1-src/unittests/test_all.py", line 183, in run_suite
    result = unittest.TextTestRunner(verbosity=2).run(create_suite())
  File "/home/teto/pygccxml-v1.9.1-src/unittests/test_all.py", line 178, in create_suite
    main_suite.addTest(tester.create_suite())
  File "/home/teto/pygccxml-v1.9.1-src/unittests/decl_printer_tester.py", line 69, in create_suite
    suite.addTest(unittest.makeSuite(Test))
  File "/nix/store/2lb6nn8ivk1alhckv43n7734lqwbw7h9-python3-3.6.2/lib/python3.6/unittest/loader.py", line 503, in makeSuite
    testCaseClass)
  File "/nix/store/2lb6nn8ivk1alhckv43n7734lqwbw7h9-python3-3.6.2/lib/python3.6/unittest/loader.py", line 92, in loadTestsFromTestCase
    loaded_suite = self.suiteClass(map(testCaseClass, testCaseNames))
  File "/nix/store/2lb6nn8ivk1alhckv43n7734lqwbw7h9-python3-3.6.2/lib/python3.6/unittest/suite.py", line 24, in __init__
    self.addTests(tests)
  File "/nix/store/2lb6nn8ivk1alhckv43n7734lqwbw7h9-python3-3.6.2/lib/python3.6/unittest/suite.py", line 57, in addTests
    for test in tests:
  File "/home/teto/pygccxml-v1.9.1-src/unittests/decl_printer_tester.py", line 42, in __init__
    compilation_mode=parser.COMPILATION_MODE.FILE_BY_FILE)
  File "/home/teto/pygccxml-v1.9.1-src/pygccxml/parser/project_reader.py", line 264, in read_files
    return self.__parse_file_by_file(files)
  File "/home/teto/pygccxml-v1.9.1-src/pygccxml/parser/project_reader.py", line 292, in __parse_file_by_file
    decls = reader.read_file(header)
  File "/home/teto/pygccxml-v1.9.1-src/pygccxml/parser/source_reader.py", line 356, in read_file
    return self.read_cpp_source_file(source_file)
  File "/home/teto/pygccxml-v1.9.1-src/pygccxml/parser/source_reader.py", line 375, in read_cpp_source_file
    xml_file = self.create_xml_file(ffname)
  File "/home/teto/pygccxml-v1.9.1-src/pygccxml/parser/source_reader.py", line 324, in create_xml_file
    ": %s status:%s" % (gccxml_msg, exit_status))
RuntimeError: Error occurred while running CASTXML:  status:1
```

so I've tried to add differents packages such as` llvm.clangStdenv clang`: which found some headers but not all, stdarg.h seems to be clang specific yet and though I have add clang, it's gcc includes that seem to be used ? 
```
python -m unittests.test_allIn file included from /home/teto/pygccxml-v1.9.1-src/unittests/data/core_types.hpp:9:
In file included from /nix/store/rww78vdn2rkayrnqsjl8ib5iq2vfm3sw-gcc-6.4.0/include/c++/6.4.0/iostream:39:
In file included from /nix/store/rww78vdn2rkayrnqsjl8ib5iq2vfm3sw-gcc-6.4.0/include/c++/6.4.0/ostream:38:
In file included from /nix/store/rww78vdn2rkayrnqsjl8ib5iq2vfm3sw-gcc-6.4.0/include/c++/6.4.0/ios:38:
In file included from /nix/store/rww78vdn2rkayrnqsjl8ib5iq2vfm3sw-gcc-6.4.0/include/c++/6.4.0/iosfwd:40:
In file included from /nix/store/rww78vdn2rkayrnqsjl8ib5iq2vfm3sw-gcc-6.4.0/include/c++/6.4.0/bits/postypes.h:40:
In file included from /nix/store/rww78vdn2rkayrnqsjl8ib5iq2vfm3sw-gcc-6.4.0/include/c++/6.4.0/cwchar:44:
/nix/store/qxb0x9fcs60zv719yy0zm95gyd8mmgks-glibc-2.25-49-dev/include/wchar.h:40:11: fatal error: 
      'stdarg.h' file not found
# include <stdarg.h>
          ^~~~~~~~~~
1 error generated.
Traceback (most recent call last):
  File "/nix/store/2lb6nn8ivk1alhckv43n7734lqwbw7h9-python3-3.6.2/lib/python3.6/runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "/nix/store/2lb6nn8ivk1alhckv43n7734lqwbw7h9-python3-3.6.2/lib/python3.6/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/home/teto/pygccxml-v1.9.1-src/unittests/test_all.py", line 193, in <module>
    sys.exit(run_suite())
  File "/home/teto/pygccxml-v1.9.1-src/unittests/test_all.py", line 183, in run_suite

```
should I set CXXFLAGS myself or can a package do it for me ?

Mic92 commented:
Getting the included path for standard c++ headers is a bit tricky. The following expression might help:

```console
$ nix-instantiate --eval --expr 'with (import <nixpkgs>) {}; clang.default_cxx_stdlib_compile'
"-isystem $(echo -n /nix/store/x3wnjni3my5r8d8vpja3qsygvg4bcbgl-gcc-6.4.0/include/c++/*) -isystem $(echo -n /nix/store/x3wnjni3my5r8d8vpja3qsygvg4bcbgl-gcc-6.4.0/include/c++/*)/$(/nix/store/x3wnjni3my5r8d8vpja3qsygvg4bcbgl-gcc-6.4.0/bin/gcc -dumpmachine)"
```

teto commented:
I changed the checkPhase to
```
   checkPhase = ''
    export CFLAGS="${clang.default_cxx_stdlib_compile}"
    echo $CFLAGS
    python -m unittests.test_all
  '';
```
but the code would still fail to find stdarg.h, I searched the echoed CFLAGS for stdarg.h
```
$ fd stdarg.h /nix/store/rww78vdn2rkayrnqsjl8ib5iq2vfm3sw-gcc-6.4.0/include/c++/6.4.0                                       
/nix/store/rww78vdn2rkayrnqsjl8ib5iq2vfm3sw-gcc-6.4.0/include/c++/6.4.0/tr1/stdarg.h
```
and it is there but in a subfolder. The homepage https://github.com/gccxml/pygccxml doesn't list clang compatibility (only python). Maybe I need to add a flag in CFLAG so that stdarg.h appears in the upper folder. Like -std=c++11 ?

teto commented:
I tried enabling pygccxml tests but it proved kinda hard, environment variables are not useful so one needs to generate a file and the include_paths is checked so that `-isystem $(echo -n /nix/store/x3wnjni3my5r8d8vpja3qsygvg4bcbgl-gcc-6.4.0/include/c++/*) -isystem $(echo -n /nix/store/x3wnjni3my5r8d8vpja3qsygvg4bcbgl-gcc-6.4.0/include/c++/*)/$(/nix/store/x3wnjni3my5r8d8vpja3qsygvg4bcbgl-gcc-6.4.0/bin/gcc -dumpmachine)` is not ok.

Meanwhile it would be cool if it could be merged without the tests as I need it for my work.

PS: Here is what I had in my attempt to enable tests (for my future self or anyone interested in taking over)
```
  xml_generator = writeText "xml_generator.cfg" ''
    [xml_generator]
    # Specify which xml generator you want to use "castxml" or "gccxml"
    # "castxml is the default"
    xml_generator=castxml
    # Path to castxml or gccxml executable file
    xml_generator_path=${castxml}/bin/castxml
    # Set the path to the compiler (for example "/usr/bin/gcc") for CastXML
    compiler_path=${llvmPackages.clang-unwrapped}/bin/clang++
    # Gccxml working directory - optional, could be set to your source code directory
    working_directory=
    # Additional include directories, as list of paths ["path1/file1.h", "path2/file2.h", ...]
    include_paths=
    # You can explicitly set what compiler it should emulate (for GCCXML)
    # Valid options are: g++, msvc6, msvc7, msvc71, msvc8, cl.
    compiler=
    # Keep xml files after errors (useful for debugging)
    keep_xml=1

    '';

  checkPhase = ''
    # export CFLAGS="-I${llvmPackages.clang-unwrapped.out}/lib/clang/4.0.1/include"
 # - if [[ $CPPSTD == "-std=c++14" ]]; then
 #      mv unittests/configs/travis_std14.cfg unittests/xml_generator.cfg;
# fi
    # echo $CFLAGS
    cp ${xml_generator} unittests/xml_generator.cfg
    python3 -m unittests.test_all
  '';
debug  checkInputs = [which sphinx llvmPackages.libcxxClang pycodestyle llvmPackages.clang-unwrapped];
```

teto committed:
pybindgen: init at 0.18.0

Helps generating python bindings to C. Required to generate
www.nsnam.org bindings.

teto commented:
@GrahamcOfBorg build pythonPackages.pybindgen pythonPackages.pygccxml castxml

GrahamcOfBorg commented:
Success on x86_64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.30473&attempt_id=fee133bc-1601-4a6e-b519-ed7a3fe9fde0)

Attempted: pythonPackages.pybindgen, pythonPackages.pygccxml, castxml

<details><summary>Partial log (click to expand)</summary><p>

```
running build_ext

----------------------------------------------------------------------
Ran 0 tests in 0.000s


OK
/nix/store/p1v8y9g1yrbab2m22h48m05wf42rhyf0-python2.7-PyBindGen-0.18.0
/nix/store/9gsqlkirlp26462bckajjy03xg4fzx06-python2.7-pygccxml-1.9.1
/nix/store/67sl6kydgifzjv37sh7kzzwv5g71cli7-CastXML-20180403
```
</p></details>

GrahamcOfBorg commented:
Success on aarch64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.30473&attempt_id=5d61f63c-e0a7-4477-989f-47309736821d)

Attempted: pythonPackages.pybindgen, pythonPackages.pygccxml, castxml

<details><summary>Partial log (click to expand)</summary><p>

```
writing manifest file 'PyBindGen.egg-info/SOURCES.txt'
running build_ext

----------------------------------------------------------------------
Ran 0 tests in 0.000s

OK
/nix/store/wrnfb9766fhc0k2bayk2bmz3n92s4lns-python2.7-PyBindGen-0.18.0
/nix/store/hqqn0rjy5la9m9939zw8dgghxbna4jv6-python2.7-pygccxml-1.9.1
/nix/store/sik3875plq6pb8vd4gkja3w4j5ig1975-CastXML-20180403
```
</p></details>

teto commented:
@GrahamcOfBorg build python3Packages.pybindgen python3Packages.pygccxml castxml

GrahamcOfBorg commented:
Success on x86_64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.30473&attempt_id=03074dbf-8bed-4ad7-8503-473d9b98cac0)

Attempted: python3Packages.pybindgen, python3Packages.pygccxml, castxml

<details><summary>Partial log (click to expand)</summary><p>

```
writing manifest file 'PyBindGen.egg-info/SOURCES.txt'
running build_ext

----------------------------------------------------------------------
Ran 0 tests in 0.000s

OK
/nix/store/vsqdazl4s4za7ijixmdfw3r0yv51sfvk-python3.6-PyBindGen-0.18.0
/nix/store/nf4zxdvs615i0gvj9r9br6gchgvxd02p-python3.6-pygccxml-1.9.1
/nix/store/67sl6kydgifzjv37sh7kzzwv5g71cli7-CastXML-20180403
```
</p></details>

GrahamcOfBorg commented:
Success on aarch64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.30473&attempt_id=a57edabf-39d4-4155-bb35-cf44e5ddf2bf)

Attempted: python3Packages.pybindgen, python3Packages.pygccxml, castxml

<details><summary>Partial log (click to expand)</summary><p>

```
writing manifest file 'PyBindGen.egg-info/SOURCES.txt'
running build_ext

----------------------------------------------------------------------
Ran 0 tests in 0.001s

OK
/nix/store/qs0nh9vczw5l2iivhs29ng86rj17d3xi-python3.6-PyBindGen-0.18.0
/nix/store/hhcxwhznf58k1cfkygq0plp4lsv6bvl3-python3.6-pygccxml-1.9.1
/nix/store/sik3875plq6pb8vd4gkja3w4j5ig1975-CastXML-20180403
```
</p></details>

teto commented:
I added pybindgen so as to support python bindings in https://github.com/NixOS/nixpkgs/pull/31346 (they released ns3.28 and fixed some problems in the build so I can finally update that PR). 
The maintainer of pygccxml clearly stated that the tests were tailored for its CI & when updating castxml, it went from 4 to 96 errors mostly because it checks the generated gcc command lines and nixos adds its own path -isystem ... that confuses the regexes. So I disabled the tests in both case.
I marked the PR as ready.

Mic92 commented:
Thanks!
